File: backend/app/routers/routing.py
import logging
from typing import Optional
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field
from app.services.graphhopper import get_route

logger = logging.getLogger("opengpx")

router = APIRouter()


class Waypoint(BaseModel):
    lat: float
    lng: float


class SegmentOptions(BaseModel):
    adventure: Optional[int] = Field(default=None, ge=0, le=100)
    avoid_motorways: bool = False
    avoid_highways: bool = False
    avoid_primary: bool = False
    prefer_unpaved: bool = False
    prefer_secondary: bool = False
    extreme: bool = False
    extreme_radius_km: float = Field(default=20.0, ge=1.0, le=300.0)
    extreme_direction: Optional[str] = None  # "N" | "S" | "E" | "O"
    extreme_loop: bool = False
    extreme_pass_index: int = Field(default=0, ge=0, le=9)


class RouteRequest(BaseModel):
    waypoints: list[Waypoint]
    adventure: int = Field(default=0, ge=0, le=100)
    segment_options: list[SegmentOptions] = []


class ElevationPoint(BaseModel):
    d: float
    ele: float


class RoadStats(BaseModel):
    road_class: dict[str, float]
    surface: dict[str, float]


class ExtremeLogEntry(BaseModel):
    segment: int
    name: str
    lat: float
    lng: float
    ele: float
    ctd_m: int
    status: str
    mode: str = ""
    passes_found: int = 0


class RouteResponse(BaseModel):
    distance_m: float
    duration_s: float
    geometry: list[list[float]]
    elevation: list[ElevationPoint]
    max_elevation: Optional[float]
    min_elevation: Optional[float]
    road_stats: RoadStats
    extreme_log: list[ExtremeLogEntry] = []


@router.post("/route", response_model=RouteResponse)
async def calculate_route(body: RouteRequest):
    if len(body.waypoints) < 2:
        raise HTTPException(status_code=400, detail="At least 2 waypoints required")

    try:
        seg_dicts = [seg.model_dump() for seg in body.segment_options]
        for i, s in enumerate(seg_dicts):
            if s.get("extreme"):
                logger.info(f"[API] seg={i} EXTREME radius={s.get('extreme_radius_km')}km dir={s.get('extreme_direction')} loop={s.get('extreme_loop')}")
        result = await get_route(
            [wp.model_dump() for wp in body.waypoints],
            body.adventure,
            seg_dicts,
        )
    except RuntimeError as exc:
        raise HTTPException(status_code=502, detail=str(exc))

    return result

File: frontend/src/api/routing.ts
import type { RouteResult, Waypoint, SegmentOptions, GlobalFilters } from '../types'

export async function fetchRoute(
  waypoints: Waypoint[],
  adventure: number,
  segmentOptions: Record<string, SegmentOptions>,
  globalFilters: GlobalFilters,
): Promise<RouteResult | null> {
  const segOpts = waypoints.slice(0, -1).map((wp) => {
    const opts = segmentOptions[wp.id]
    return {
      adventure: opts?.adventure ?? null,
      avoid_motorways: (opts?.avoid_motorways || globalFilters.avoid_motorways),
      avoid_highways: (opts?.avoid_highways || globalFilters.avoid_highways),
      avoid_primary: (opts?.avoid_primary || globalFilters.avoid_primary),
      prefer_unpaved: (opts?.prefer_unpaved || globalFilters.prefer_unpaved),
      prefer_secondary: (opts?.prefer_secondary || globalFilters.prefer_secondary),
      extreme: (opts?.extreme || false),
      extreme_radius_km: opts?.extreme_radius_km ?? 20,
      extreme_direction: opts?.extreme_direction ?? null,
      extreme_loop: opts?.extreme_loop ?? false,
      extreme_pass_index: opts?.extreme_pass_index ?? 0,
    }
  })

  const resp = await fetch('/api/route', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      waypoints: waypoints.map((w) => ({ lat: w.lat, lng: w.lng })),
      adventure,
      segment_options: segOpts,
    }),
  })

  if (!resp.ok) {
    const err = await resp.json().catch(() => ({ detail: 'Unknown error' }))
    throw new Error(err.detail ?? `HTTP ${resp.status}`)
  }

  return resp.json()
}
